pageserver/src/storage_sync.rs

```diff
@@ -129,1441 +114,660 @@
-    };
-    sync_queue.push(
-        TenantTimelineId {
-            tenant_id,
-            timeline_id,
-        },
-        SyncTask::delete(LayersDeletion {
-            layers_to_delete,
-            deleted_layers: HashSet::new(),
-            deletion_registered: false,
-        }),
-    );
-    debug!("Deletion task for tenant {tenant_id}, timeline {timeline_id} sent")
-}
+    /// Like init_upload_queue, but the list of files is initially empty, and
+    /// metadata is passed as a TimelineMetadata. This is used when creating a
+    /// new timeline.
+    pub fn init_upload_queue_empty(&self, metadata: &TimelineMetadata) -> anyhow::Result<()> {
+        info!(
+            "initializing upload queue for new timeline with disk_consistent_lsn: {}",
+            metadata.disk_consistent_lsn()
+        );
 
-/// Requests the download of the entire timeline for a given tenant.
-/// No existing local files are currently overwritten, except the metadata file (if its disk_consistent_lsn is less than the downloaded one).
-/// The metadata file is always updated last, to avoid inconsistencies.
-///
-/// On any failure, the task gets retried, omitting already downloaded layers.
-///
-/// Ensure that the loop is started otherwise the task is never processed.
-pub fn schedule_layer_download(tenant_id: TenantId, timeline_id: TimelineId) {
-    debug!("Scheduling layer download for tenant {tenant_id}, timeline {timeline_id}");
-    let sync_queue = match SYNC_QUEUE.get() {
-        Some(queue) => queue,
-        None => {
-            warn!("Could not send download task for tenant {tenant_id}, timeline {timeline_id}");
-            return;
+        let mut upload_queue = self.upload_queue.lock().unwrap();
+        if upload_queue.latest_metadata.is_some()
+            || upload_queue.last_uploaded_consistent_lsn.is_some()
+        {
+            anyhow::bail!("attempted to initialize already initialized upload queue")
         }
-    };
-    sync_queue.push(
-        TenantTimelineId {
-            tenant_id,
-            timeline_id,
-        },
-        SyncTask::download(LayersDownload::from_skipped_layers(HashSet::new())),
-    );
-    debug!("Download task for tenant {tenant_id}, timeline {timeline_id} sent")
-}
 
-/// Local existing timeline files
-///
-/// Values of this type serve different meanings in different contexts. On startup, collected
-/// timelines come with the full collected information and when signalling readyness to attach
-/// after completed download. After the download the file information is no longer carried, because
-/// it is already merged into [`RemoteTimeline`].
-#[derive(Debug)]
-pub struct TimelineLocalFiles(TimelineMetadata, HashMap<PathBuf, LayerFileMetadata>);
-
-impl TimelineLocalFiles {
-    pub fn metadata(&self) -> &TimelineMetadata {
-        &self.0
+        upload_queue.latest_metadata = Some(metadata.clone());
+        upload_queue.last_uploaded_consistent_lsn = Some(metadata.disk_consistent_lsn());
```

Comment:
I think this is wrong. These values are indeed the latest ones. But data they reference may not be uploaded at the moment of this assignment. So this will give safekeepers remote conistent lsn that is not in fact consistent with the remote.

The same problem exists for schedule_index_upload. We cannot update these values before actual upload finishes

Reply:
I'll look into this

Reply:
So, this initialization is definitely wrong.
But I don't think it's wrong in `schedule_index_upload`:
`latest_metadata` is documented to be ahead, and I think it's fine because index upload currently acts like a barrier.
`last_uploaded_consistent_lsn` is only updated once the index upload operation completes.

Will create a PR with a patch that cleans up some things and adds comments.